doc/dev/sharp_bits.rst

```diff
@@ -0,0 +1,1015 @@
+
+With the quantum function qjit-compiled, the optimization loop
+now takes 16.4s ± 1.51s.
+
+However, while the quantum function is now compiled, and the compiled function
+is called to compute cost and gradient values, the optimization loop is still
+occuring in Python.
+
+Instead, we can write the optimization loop itself as a function and decorate
+it with ``@qjit``; this will compile the optimization loop, and allow the full
+optimization to take place within Catalyst:
+
+.. code-block:: python
+
+    @qjit
+    def optimize(init_weights, data, steps):
+        def loss(x):
+            dy = grad(cost, argnum=0)(x, data)[0]
+            return (cost(x, data), dy)
+
+        opt = jaxopt.GradientDescent(loss, stepsize=0.4, value_and_grad=True)
+        update_step = lambda i, *args: tuple(opt.update(*args))
+
+        params = init_weights
+        state = opt.init_state(params)
+
+        return for_loop(0, steps, 1)(update_step)(params, state)[0]
+
+The optimization now takes 574ms ± 43.1ms to complete when using 200 steps.
+Note that, to compute gradients within a qjit-compiled function,
+the :func:`catalyst.grad` function must be used.
+
+JAX support and restrictions
+----------------------------
+
+Catalyst is utilizes JAX for program capture, which means you are able to
+leverage the many functions accessible in ``jax`` and ``jax.numpy`` to write
+code that supports :func:`@qjit <~.qjit>` and dynamic variables.
+
+Currently, we are aiming to support as many JAX functions as possible, however
+there may be cases where there is missing coverage. Known JAX functionality
+that doesn't work with Catalyst includes:
+
+- ``jax.numpy.polyfit``
+- ``jax.numpy.fft``
+- ``jax.debug``
+- ``jax.numpy.ndarray.at[index]`` when ``index`` corresponds to all array
+  indices.
+
+If you come across any other JAX functions that don't work with Catalyst
+(or don't already have a Catalyst equivalents), please let us know by opening
+a `GitHub issue <https://github.com/PennyLaneAI/catalyst/issues>`__.
+
+While leveraging ``jax.numpy`` makes it easy to port over NumPy-based
+PennyLane workflows to Catalyst, we also inherit `various restrictions
+and 'gotchas' from JAX
+<https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html>`__.
+This includes:
+
+* **Pure functions**: Compilation is primarily designed to only work on pure
+  functions. That is, functions that do not have any side-effects; the
+  output is purely dependent only on function inputs.
+
+* **In-place array updates**: Rather than using in-place array updates, the
+  syntax ``new_array = jax_array.at[index].set(value)`` should be used. For
+  more details, see `jax.numpy.ndarray.at
+  <https://jax.readthedocs.io/en/latest/_autosummary /jax.numpy.ndarray.at.html>`__.
+
+* **Lack of stateful random number generators**: In JAX, random number
+  generators need to be explicitly created within the :func:`@qjit <~.qjit>` function
+  using ``jax.random.PRNGKey(int)``:
```

Comment:
I don't think the `within the qjit function` part is accurate, the key can also both, be created as a constant outside the function (closure), and passed in as an argument (a JAX prng key is just a regular array).

I think what might also be important to show here is the "stateless" nature of the PRNG, so the key state must be explicitly updated each time you want to compute a random number:
```py
key = jax.random.PRNGKey(seed)

key, subkey = jax.random.split(key)
r1 = jax.random.normal(subkey, shape=(1,))

key, subkey = jax.random.split(key)
r2 = jax.random.normal(subkey, shape=(1,))
```

It would also be good to link to the JAX docs here, maybe
https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html
or
https://jax.readthedocs.io/en/latest/jax.random.html

Reply:
```suggestion
* **Lack of stateful random number generators**: In JAX, random number
  generators are stateless, and the key state must be explicitly updated each time you want to compute a random number. For more details, see the `JAX documentation <https://jax.readthedocs.io/en/latest/jax-101/05-random-numbers.html>`__.
```